### PR TITLE
server: extend tag completion

### LINKED ISSUE #17: Autocomplete tags with prefixes
Prefixes are not accounted for right now. Need to both accept prefixed tags for completion and return prefixed results.

### LINKED ISSUE #25: Autocomplete system tags

### PR DESCRIPTION
**Pull request by Kamuso**
Closes #17, closes #25
Left out #54 because we pretty much have the "all categories" functionality by just not specifying a category, and undefined is used for regular tags.

### CODE REVIEW

File: db/tags.go [hunk 1]
```diff
@@ -96,17 +98,98 @@ func RemoveTags(imageID int64, tags []common.Tag) error {
...
-	// TODO: Inject system tags into array
+	i := strings.IndexByte(s, ':')
+	if i != -1 {
+		// Complete postfix
+		re := regexp.MustCompile(`^` + regexp.QuoteMeta(s[i+1:]) + `.*`)
+		switch s[:i] {
+		case "undefined":
+			prefix += s[:i+1]
+			s = s[i+1:]
+			typeQ = fmt.Sprintf(" and type = %d", common.Undefined)
```
Comment (bakape) [inline db/tags.go]: Please move all cases of this into an embedded function to DRY this a bit. 
Something like 
```
completeWithPrefix := func (t common.TagType) {
  prefix += s[:i+1]
  s = s[i+1:]
  typeQ = fmt.Sprintf(" and type = %d", t)
}
```

File: db/tags.go [hunk 2]
```diff
@@ -96,17 +98,98 @@ func RemoveTags(imageID int64, tags []common.Tag) error {
...
+	if s[0] == '-' {
+		prefix = "-"
+		s = s[1:]
+	}
 
-	// TODO: Inject system tags into array
+	i := strings.IndexByte(s, ':')
+	if i != -1 {
+		// Complete postfix
+		re := regexp.MustCompile(`^` + regexp.QuoteMeta(s[i+1:]) + `.*`)
```
Comment (bakape) [inline db/tags.go]: Return error instead of panicking.

File: db/tags.go [hunk 3]
```diff
@@ -96,17 +98,98 @@ func RemoveTags(imageID int64, tags []common.Tag) error {
...
+		case "system":
+			tags = matchPost(re, "system:",
+				[]string{"size", "width", "height", "duration", "tag_count"})
+			return
+		case "order":
+			prefix = s[:i+1]
+			if s[i+1:] != "" {
+				if s[i+1] == '-' {
+					prefix = "order:-"
+					re = regexp.MustCompile(`^` + regexp.QuoteMeta(s[i+2:]) + `.*`)
```
Comment (bakape) [inline db/tags.go]: Return error instead of panicking.

File: db/tags.go [hunk 4]
```diff
@@ -96,17 +98,98 @@ func RemoveTags(imageID int64, tags []common.Tag) error {
...
+				[]string{"size", "width", "height", "duration", "tag_count", "random"})
+			return
+		case "limit":
+			return
+		default:
+			// Continue as regular tag
+		}
+	} else {
+		// Complete prefix
+		re := regexp.MustCompile(`^` + regexp.QuoteMeta(s) + `.*`)
```
Comment (bakape) [inline db/tags.go]: Return error instead of panicking.

**Review by bakape — CHANGES_REQUESTED** [review]

**Comment by Kamuso** [comment]
Ok, updated. So what about #54, should it be closed?

**Comment by Kamuso** [comment]
While I'm here, fixed a bug.

**Comment by bakape** [comment]
Okay, thank you!

**Review by bakape — APPROVED** [review]

**Pull request merged**